zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/MultiResourceDeploymentTest.java

```diff
@@ -0,0 +1,345 @@
+                  DecisionIntent.CREATED,
+                  FormIntent.CREATED)
+              .limit(4)
+              .toList();
+
+      // when
+      final var secondDeployment =
+          engine
+              .deployment()
+              // deploy the exact same resources again
+              .withXmlResource(PROCESS_V1)
+              .withXmlClasspathResource(DMN_V1)
+              .withJsonClasspathResource(FORM_V1)
+              .deploy()
+              .getValue();
+
+      // then
+      assertThat(secondDeployment.getProcessesMetadata())
+          .singleElement()
+          .satisfies(
+              metadata ->
+                  Assertions.assertThat(metadata)
+                      .hasVersion(1)
+                      .isDuplicate()
+                      .extracting(
+                          ProcessMetadataValue::getProcessDefinitionKey,
+                          InstanceOfAssertFactories.LONG)
+                      .isEqualTo(processV1.getProcessDefinitionKey()));
+      assertThat(secondDeployment.getDecisionRequirementsMetadata())
+          .singleElement()
+          .satisfies(
+              metadata ->
+                  Assertions.assertThat(metadata)
+                      .hasDecisionRequirementsVersion(1)
+                      .isDuplicate()
+                      .extracting(
+                          DecisionRequirementsMetadataValue::getDecisionRequirementsKey,
+                          InstanceOfAssertFactories.LONG)
+                      .isEqualTo(drgV1.getDecisionRequirementsKey()));
+      assertThat(secondDeployment.getDecisionsMetadata())
+          .singleElement()
+          .satisfies(
+              metadata ->
+                  Assertions.assertThat(metadata)
+                      .hasVersion(1)
+                      .isDuplicate()
+                      .extracting(
+                          DecisionRecordValue::getDecisionKey,
+                          DecisionRecordValue::getDecisionRequirementsKey)
+                      .containsExactly(
+                          decisionV1.getDecisionKey(), drgV1.getDecisionRequirementsKey()));
+      assertThat(secondDeployment.getFormMetadata())
+          .singleElement()
+          .satisfies(
+              metadata ->
+                  Assertions.assertThat(metadata)
+                      .hasVersion(1)
+                      .isDuplicate()
+                      .extracting(FormMetadataValue::getFormKey)
+                      .isEqualTo(formV1.getFormKey()));
+
+      final var recordsAfterSecondDeployment =
+          RecordingExporter.records()
+              .withIntents(
+                  ProcessIntent.CREATED,
+                  DecisionRequirementsIntent.CREATED,
+                  DecisionIntent.CREATED,
+                  FormIntent.CREATED)
+              .toList();
+      assertThat(recordsAfterSecondDeployment)
+          .containsExactlyElementsOf(recordsAfterFirstDeployment);
+    }
```

Comment:
I think this can be more performant, but it's up to you if you like how it looks.
```suggestion
    public void shouldNotCreateNewVersionsIfNoResourceHasChanged() {
      // given
      final var firstDeployment =
          engine
              .deployment()
              .withXmlResource(PROCESS_V1)
              .withXmlClasspathResource(DMN_V1)
              .withJsonClasspathResource(FORM_V1)
              .deploy();
      final var firstDeploymentValue = firstDeployment.getValue();
      final var processV1 = firstDeploymentValue.getProcessesMetadata().getFirst();
      final var drgV1 = firstDeploymentValue.getDecisionRequirementsMetadata().getFirst();
      final var decisionV1 = firstDeploymentValue.getDecisionsMetadata().getFirst();
      final var formV1 = firstDeploymentValue.getFormMetadata().getFirst();
      final long recordsCountBefore =
          RecordingExporter.records()
              .onlyEvents()
              .limit(l -> l.getPosition() >= firstDeployment.getPosition())
              .count();
      assertThat(recordsCountBefore)
          .isEqualTo(5)
          .describedAs("Should have event records for 4 resources and 1 for deployment created");
      // when
      final Record<DeploymentRecordValue> secondDeployment =
          engine
              .deployment()
              // deploy the exact same resources again
              .withXmlResource(PROCESS_V1)
              .withXmlClasspathResource(DMN_V1)
              .withJsonClasspathResource(FORM_V1)
              .deploy();
      final var secondDeploymentValue = secondDeployment.getValue();
      // then
      assertThat(secondDeploymentValue.getProcessesMetadata())
          .singleElement()
          .satisfies(
              metadata ->
                  Assertions.assertThat(metadata)
                      .hasVersion(1)
                      .isDuplicate()
                      .extracting(
                          ProcessMetadataValue::getProcessDefinitionKey,
                          InstanceOfAssertFactories.LONG)
                      .isEqualTo(processV1.getProcessDefinitionKey()));
      assertThat(secondDeploymentValue.getDecisionRequirementsMetadata())
          .singleElement()
          .satisfies(
              metadata ->
                  Assertions.assertThat(metadata)
                      .hasDecisionRequirementsVersion(1)
                      .isDuplicate()
                      .extracting(
                          DecisionRequirementsMetadataValue::getDecisionRequirementsKey,
                          InstanceOfAssertFactories.LONG)
                      .isEqualTo(drgV1.getDecisionRequirementsKey()));
      assertThat(secondDeploymentValue.getDecisionsMetadata())
          .singleElement()
          .satisfies(
              metadata ->
                  Assertions.assertThat(metadata)
                      .hasVersion(1)
                      .isDuplicate()
                      .extracting(
                          DecisionRecordValue::getDecisionKey,
                          DecisionRecordValue::getDecisionRequirementsKey)
                      .containsExactly(
                          decisionV1.getDecisionKey(), drgV1.getDecisionRequirementsKey()));
      assertThat(secondDeploymentValue.getFormMetadata())
          .singleElement()
          .satisfies(
              metadata ->
                  Assertions.assertThat(metadata)
                      .hasVersion(1)
                      .isDuplicate()
                      .extracting(FormMetadataValue::getFormKey)
                      .isEqualTo(formV1.getFormKey()));

      final long recordsCountAfter =
          RecordingExporter.records()
              .onlyEvents()
              .limit(l -> l.getPosition() >= secondDeployment.getPosition())
              .count();
      assertThat(recordsCountAfter - recordsCountBefore)
          .isEqualTo(1)
          .describedAs("Should only have one extra deployment created");
    }```

Reply:
Thanks, @mustafadagher! That's indeed even better 🚀.